salt/modules/opkg.py

```diff
@@ -370,15 +374,172 @@ def _parse_reported_packages_from_install_output(output):
+        if len(line_tokens) >= 3 and line_tokens[2].startswith('('):
+            package = line_tokens[1]
+    
+    return package
+
+
+def _get_operation_and_package_from_output_line(line):
+    '''
+    Returns the (operation, package-name) combination if the line is a valid opkg operation output
+    If operation or package-name could not be parsed from the line, it returns None instead of the
+    operation/package-name
+    '''
+    operation = _get_operation_from_output_line(line)
+    current_package = _get_package_from_output_line(operation, line)
+
+    return operation, current_package
+
+
+def _process_with_progress(cmd, jid, total_packages_count):
+    '''
+    Process the opkg command and fire events to the salt-master bus
+    indicating the progress of the operation
+    '''
+    out = {}
+    stdout = ''
+    minion_id = str(salt.config.get_id(__opts__)[0])
+    notify_progress_period = __opts__.get('pkg_progress_period', 60)
+    last_notify_timestamp = 0
+    processed_count = 0
+    force_update = True
+    last_sent_package = None
+    current_package = None
+    operation = None
+    event = salt.utils.event.get_event(
+            'minion', opts=__opts__, listen=False
+            )
+    tag = 'salt/minion/pkg_progress/{0}'.format(minion_id)
+    proc = subprocess.Popen(
+            cmd,
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            bufsize=1,
+            universal_newlines=True
+            )
+    # We don't want the standard output to be blocking to avoid missing updates
+    # based on specified progress period
+    try:
+        flags = fcntl(proc.stdout, F_GETFL)
+        stdout_fileno = proc.stdout.fileno()
+        fcntl(stdout_fileno, F_SETFL, flags | os.O_NONBLOCK)
+        while True:
+            ready_to_read = select.select([stdout_fileno], [], [])[0]
+            if ready_to_read:
+                output = proc.stdout.readline()
+            else:
+                output = ''
+                time.sleep(1)
+            if output == '' and proc.poll() is not None:
+                break
+            if output:
+                stdout += output
+                line_operation, line_package = _get_operation_and_package_from_output_line(output)
+                if not line_operation or not line_package:
+                    continue
+
+                operation = line_operation
+                current_package = line_package
+                if operation in ['install', 'upgrade', 'remove', 'downgrade']:
+                    processed_count += 1
+                    force_update |= processed_count in [1, total_packages_count]
+            timestamp = time.time()
+            if (timestamp - last_notify_timestamp > notify_progress_period or force_update) and \
+                    last_sent_package != current_package:
+                last_sent_package = current_package
```

Comment:
I think you don't need the last_sent_package logic. In my understanding the read will block until you get a new line which will be when it installs/removes/upgrades/downgrades a new package. So current_package will always be different from last_sent_package, with probably one small exception: when you have lines that are not operations. But in that case you anyway don't send a notification because you circuit-break the logic.